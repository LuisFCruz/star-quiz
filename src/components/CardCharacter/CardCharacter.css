
.CardCharacter-card {
    max-width: 400;
}

.CardCharacter-media {
  height: 200px;
  width: 100%;
}

.CardCharacter-actions {
  justify-content: center;
  flex-wrap: wrap
}

.CardCharacter-buttons {
  justify-content: space-around;
  display: flex;
  width: 100%;
}
