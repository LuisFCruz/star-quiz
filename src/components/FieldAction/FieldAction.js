import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { TextField, withStyles, Button } from '@material-ui/core';

const styles = {
  container: {
    display: 'flex',
    alignItems: 'inherit',
  },
  button: {
    marginLeft: '5px',
    minWidth: 'auto',
  }
};

export class FieldAction extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    onConfirm: PropTypes.func.isRequired,
    placeholder: PropTypes.string,
  }

  constructor() {
    super();

    this.state = {
      value: ''
    };
  }

  handleConfirmClick = () => {
    const { value } = this.state;
    this.props.onConfirm(value);
  }

  handleChange = ({target: {value = ''}}) => {
    this.setState({ value });
  }

  render() {
    const { classes, placeholder = '' } = this.props;
    return (
      <div className={classes.container}>
        <TextField
          placeholder={placeholder}
          onChange={this.handleChange}
        />
        <Button
          color="primary"
          variant="contained"
          className={classes.button}
          onClick={this.handleConfirmClick}
        >ok</Button>
      </div>
    )
  }
}

export default withStyles(styles)(FieldAction);
