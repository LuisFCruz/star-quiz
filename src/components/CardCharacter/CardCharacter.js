import { Button, Card, CardActions } from '@material-ui/core';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { selectCharacter, updateCharacters, sumScore } from '../../actions';
import Avatar from '../Avatar/Avatar';
import FieldAction from '../FieldAction/FieldAction';

import './CardCharacter.css';

export class CardCharacter extends Component {
  constructor() {
    super();

    this.state = {
      reply: false,
    };
  }
  static propTypes = {
    character: PropTypes.object.isRequired,
    finished: PropTypes.bool,
    updateCharacters: PropTypes.func,
    selectCharacter: PropTypes.func,
    sumScore: PropTypes.func,
  };

  handleHelpClick = () => {
    const { character } = this.props;
    this.props.updateCharacters({ ...character, helped: true });
    this.props.selectCharacter(character);
  }

  handleReplyClick = () => {
    this.setState({ reply: true });
  }
  
  handleConfirm = (name) => {
    const { character } = this.props;
    if (name && character.name.toUpperCase() === name.toUpperCase()) {
      this.props.updateCharacters({ ...character, answered: true });
      this.props.sumScore(character);
    }
    this.setState({ reply: false });
  }

  renderActions = () => {
    const { character, finished } = this.props;
    const { reply } = this.state;

    if (!finished && !character.answered && reply) {
      return <FieldAction onConfirm={this.handleConfirm} />;
    }

    return (
      <div className="CardCharacter-buttons">
        <Button
          color="primary"
          variant="contained"
          onClick={this.handleReplyClick}
          disabled={finished || character.answered}
        > ? </Button>
        <Button
          variant="contained"
          onClick={this.handleHelpClick}
          disabled={finished || character.answered}
        > ... </Button>
      </div>
    );

    
  }

  render() {
    const { character } = this.props;
    
    return (

      <Card className="CardCharacter-card">
        <Avatar id={character.id} className="CardCharacter-media"/>
        <CardActions className="CardCharacter-actions">
          {this.renderActions()}
        </CardActions>
      </Card >
    );
  }
}

const mapStateToProps = (state) => {
  const { finished } = state;
  return { finished };
};

export default connect(
  mapStateToProps,
  { selectCharacter, updateCharacters, sumScore }
)(CardCharacter);
