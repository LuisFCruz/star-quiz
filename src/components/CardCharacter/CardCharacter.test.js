import { Button } from '@material-ui/core';
import { shallow } from 'enzyme';
import React from 'react';

import { CardCharacter } from './CardCharacter';
import { FieldAction } from '../FieldAction/FieldAction';

const character = {
  answered: false,
  films: 'name',
  hairColor: 'n/a',
  height: '187',
  helped: false,
  homeworld: 'name',
  id: '1',
  name: 'r2-d2',
  species: 'name',
  vehicles: 'n/a'
};

describe('<CardCharacter />', () => {
  let wrapper;
  let component;
  
  beforeEach(() => {
    const props = {
      character,
      finished: false,
      updateCharacters: jest.fn(),
      selectCharacter: jest.fn(),
      sumScore: jest.fn()
    };
    wrapper = shallow(<CardCharacter {...props} />);
    component = wrapper.instance();
  });

  test('smoke test', () => {
    expect(CardCharacter).toBeDefined();
    expect(component).toBeInstanceOf(CardCharacter);
  });

  test('render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should have call handleReplyClick()', () => {
    const handleSpy = jest.fn();
    component.handleReplyClick = handleSpy;
    component.forceUpdate();

    const button = wrapper.find(Button);

    expect(button).toHaveLength(2);
    button.first().simulate('click');
    expect(handleSpy).toHaveBeenCalled();
  });

  test('should have changed reply value', () => {
    component.handleReplyClick();
    expect(wrapper.state('reply')).toBeTruthy();
  });

  test('should have call handleHelpClick()', () => {
    const handleSpy = jest.fn();
    component.handleHelpClick = handleSpy;
    component.forceUpdate();

    const button = wrapper.find(Button);

    expect(button).toHaveLength(2);
    button.last().simulate('click');
    expect(handleSpy).toHaveBeenCalled();
  });

  test('should have call updateCharacters() and selectCharacter()', () => {
    component.handleHelpClick();
    expect(component.props.updateCharacters).toHaveBeenCalledTimes(1);
    expect(component.props.selectCharacter).toHaveBeenCalledTimes(1);
  });

  test('should have call updateCharacters() and sumScore()', () => {
    component.handleConfirm(character.name);
    expect(component.props.updateCharacters).toHaveBeenCalledTimes(1);
    expect(component.props.sumScore).toHaveBeenCalledTimes(1);
  });

  test('should not have call updateCharacters() and sumScore()', () => {
    component.handleConfirm();
    expect(component.props.updateCharacters).toHaveBeenCalledTimes(0);
    expect(component.props.sumScore).toHaveBeenCalledTimes(0);
    expect(wrapper.state('reply')).toBeFalsy();
  });
});
