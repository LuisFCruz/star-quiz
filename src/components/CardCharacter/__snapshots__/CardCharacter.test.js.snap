// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardCharacter /> render 1`] = `
<WithStyles(Card)
  className="CardCharacter-card"
>
  <Avatar
    className="CardCharacter-media"
    id="1"
  />
  <WithStyles(CardActions)
    className="CardCharacter-actions"
  >
    <div
      className="CardCharacter-buttons"
    >
      <WithStyles(Button)
        color="primary"
        disabled={false}
        onClick={[Function]}
        variant="contained"
      >
         ? 
      </WithStyles(Button)>
      <WithStyles(Button)
        disabled={false}
        onClick={[Function]}
        variant="contained"
      >
         ... 
      </WithStyles(Button)>
    </div>
  </WithStyles(CardActions)>
</WithStyles(Card)>
`;
